File: clients/swift/Tests/QuidnugTests/MerkleTests.swift
import XCTest
import CryptoKit
@testable import Quidnug

final class MerkleTests: XCTestCase {
    private func sh(_ d: Data) -> Data { Data(SHA256.hash(data: d)) }
    private func concat(_ a: Data, _ b: Data) -> Data { var r = a; r.append(b); return r }
    private func toHex(_ d: Data) -> String { d.map { String(format: "%02x", $0) }.joined() }

    func testSingleSiblingRight() throws {
        let tx = Data("tx-1".utf8)
        let sib = sh(Data("tx-2".utf8))
        let leaf = sh(tx)
        let root = sh(concat(leaf, sib))
        let frames = [MerkleProofFrame(hash: toHex(sib), side: "right")]
        XCTAssertTrue(try Merkle.verifyInclusionProof(
            txBytes: tx, frames: frames, expectedRootHex: toHex(root)))
    }

    func testFourLeafTree() throws {
        let leaves = (0..<4).map { sh(Data("tx-\($0)".utf8)) }
        let p0 = sh(concat(leaves[0], leaves[1]))
        let p1 = sh(concat(leaves[2], leaves[3]))
        let root = sh(concat(p0, p1))
        let frames = [
            MerkleProofFrame(hash: toHex(leaves[3]), side: "right"),
            MerkleProofFrame(hash: toHex(p0), side: "left")
        ]
        XCTAssertTrue(try Merkle.verifyInclusionProof(
            txBytes: Data("tx-2".utf8), frames: frames, expectedRootHex: toHex(root)))
    }

    func testTamperedTxRejected() throws {
        let sib = sh(Data("tx-2".utf8))
        let leaf = sh(Data("tx-1".utf8))
        let root = sh(concat(leaf, sib))
        let frames = [MerkleProofFrame(hash: toHex(sib), side: "right")]
        XCTAssertFalse(try Merkle.verifyInclusionProof(
            txBytes: Data("tampered".utf8), frames: frames, expectedRootHex: toHex(root)))
    }

    func testMalformedFrameThrows() {
        let frames = [MerkleProofFrame(hash: "nothex", side: "right")]
        XCTAssertThrowsError(try Merkle.verifyInclusionProof(
            txBytes: Data("x".utf8), frames: frames,
            expectedRootHex: String(repeating: "a", count: 64)))
    }

    func testEmptyTxThrows() {
        XCTAssertThrowsError(try Merkle.verifyInclusionProof(
            txBytes: Data(), frames: [], expectedRootHex: String(repeating: "a", count: 64)))
    }
}

File: clients/dotnet/tests/MerkleTests.cs
using System.Security.Cryptography;
using System.Text;
using Xunit;

namespace Quidnug.Client.Tests;

public class MerkleTests
{
    private static byte[] Sha(byte[] data) => SHA256.HashData(data);

    [Fact]
    public void SingleSiblingRight()
    {
        byte[] tx = Encoding.UTF8.GetBytes("tx-1");
        byte[] sib = Sha(Encoding.UTF8.GetBytes("tx-2"));
        byte[] leaf = Sha(tx);
        byte[] concat = new byte[64];
        Buffer.BlockCopy(leaf, 0, concat, 0, 32);
        Buffer.BlockCopy(sib, 0, concat, 32, 32);
        byte[] root = Sha(concat);

        var frames = new[] { new Merkle.MerkleProofFrame(Convert.ToHexString(sib).ToLowerInvariant(), "right") };
        Assert.True(Merkle.VerifyInclusionProof(tx, frames, Convert.ToHexString(root).ToLowerInvariant()));
    }

    [Fact]
    public void FourLeafTree()
    {
        byte[][] leaves = new byte[4][];
        for (int i = 0; i < 4; i++) leaves[i] = Sha(Encoding.UTF8.GetBytes($"tx-{i}"));

        byte[] p0 = Sha(Concat(leaves[0], leaves[1]));
        byte[] p1 = Sha(Concat(leaves[2], leaves[3]));
        byte[] root = Sha(Concat(p0, p1));

        var frames = new[]
        {
            new Merkle.MerkleProofFrame(Convert.ToHexString(leaves[3]).ToLowerInvariant(), "right"),
            new Merkle.MerkleProofFrame(Convert.ToHexString(p0).ToLowerInvariant(), "left"),
        };
        Assert.True(Merkle.VerifyInclusionProof(
            Encoding.UTF8.GetBytes("tx-2"), frames, Convert.ToHexString(root).ToLowerInvariant()));
    }

    [Fact]
    public void TamperedTxRejected()
    {
        byte[] sib = Sha(Encoding.UTF8.GetBytes("tx-2"));
        byte[] leaf = Sha(Encoding.UTF8.GetBytes("tx-1"));
        byte[] root = Sha(Concat(leaf, sib));
        var frames = new[] { new Merkle.MerkleProofFrame(Convert.ToHexString(sib).ToLowerInvariant(), "right") };
        Assert.False(Merkle.VerifyInclusionProof(
            Encoding.UTF8.GetBytes("tampered"), frames, Convert.ToHexString(root).ToLowerInvariant()));
    }

    [Fact]
    public void MalformedFrameThrows()
    {
        var frames = new[] { new Merkle.MerkleProofFrame("nothex", "right") };
        Assert.Throws<QuidnugValidationException>(
            () => Merkle.VerifyInclusionProof(
                Encoding.UTF8.GetBytes("x"), frames, new string('a', 64)));
    }

    [Fact]
    public void EmptyTxThrows()
    {
        Assert.Throws<QuidnugCryptoException>(
            () => Merkle.VerifyInclusionProof(
                ReadOnlySpan<byte>.Empty, Array.Empty<Merkle.MerkleProofFrame>(), new string('a', 64)));
    }

    private static byte[] Concat(byte[] a, byte[] b)
    {
        byte[] c = new byte[a.Length + b.Length];
        Buffer.BlockCopy(a, 0, c, 0, a.Length);
        Buffer.BlockCopy(b, 0, c, a.Length, b.Length);
        return c;
    }
}
